File: Python/pre_order.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
'''
@File: pre_order.py
@Funciton Description: N 叉树的前序遍历
@LeetCode Link: https://leetcode.cn/problems/n-ary-tree-preorder-traversal/
@Python Version: 3.10.4
@Platform: PyTorch 1.12.1+cu113
@Author: Wei Li
@Contact: weili_yzzcq@163.com
@Time: 2022/09/28 17:50:41
@Version: 0.1
@License: Apache License Version 2.0, January 2004
'''

# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution:
    def preorder(self, root):
        answer = []
        def dfs(node: 'Node'):
            if node is None:
                return
            answer.append(node.val)
            for ch in node.children:
                dfs(ch)
        dfs(root)

        return answer


# ----------------------------
if  __name__ == "__main__":
    pass

File: C++/pre_order.cpp
/**
 * @file pre_order.cpp
 * @author Wei Li (weili_yzzcq@163.com)
 * @brief N 叉树(二叉树)的前序遍历
 * @LeetCode Link https://leetcode.cn/problems/n-ary-tree-preorder-traversal/
 * @version 0.1
 * @date 2022-09-28
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#include <vector>


// Definition for a Node.
class Node {
public:
    int val;
    std::vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, std::vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};


class Solution 
{
public:
    void helper(const Node* root, std::vector<int> & result)
    {
        if (root == nullptr)
        {
            return;
        }
        result.emplace_back(root->val);
        for (auto &ch : root->children)
        {
            helper(ch, result);
        }
    }

    std::vector<int> preorder(Node* root) 
    {
        std::vector<int> result;
        helper(root, result);

        return result;
    }
};


// -------------------------------------
int main(int argc, char const *argv[])
{
    /* code */
    return 0;
}
